src/Json.mli

```diff
@@ -99,18 +99,314 @@ let _ =
     |} 
   )
 ]}
-*) 
+*)
 
-module Decode = Json_decode
-module Encode = Json_encode
+type t = Js.Json.t
+(** The type of a JSON data structure *)
 
-exception ParseError of string
+type json = t
+(** Defined for convenience. *)
 
-val parse: string -> Js.Json.t option
-(** [parse s] returns [Some json] if s is a valid json string, [None] otherwise *)
+val to_string : json -> string
+(** JSON can be encoded as a string. *)
 
-val parseOrRaise: string -> Js.Json.t
-(** [parse s] returns a [Js.Json.t] if s is a valid json string, raises [ParseError] otherwise *)
+type exn +=
+  | Of_string_error of string
+        (** The exception raised when parsing JSON error occurs *)
 
-val stringify: Js.Json.t -> string
-(** [stringify json] returns the [string] representation of the given [Js.Json.t] value *)
+val of_string : string -> json
+(** JSON can be parsed from a string. Raises {Of_string_error}. *)
+
+type 'a to_json = 'a -> json
+(** Describe how to encode a value into JSON. *)
+
+val to_json : json to_json
+(** JSON can be encoded as JSON, trivially. *)
+
+(** The type of a error which occurs during decoding JSON values. *)
+type of_json_error = Json_error of string | Unexpected_variant of string
```

Comment:
`Unexpected_variant` currently always sends "unexpected variant" in the string payload. Should this payload be removed?

Reply:
I think I've wanted to include the tag name there.